xmtp/README.md

```diff
@@ -0,0 +1,104 @@
+# XMTP Crate
+
+**⚠️ Experimental:** Early development stage, expect frequent changes and unresolved issues.
+
+## State Machine
+
+### Simplifying assumptions
+
+- We use a pull-based approach for detecting if the installations in a conversation has changed - refreshing the installation list whenever the last refresh was more than THRESHOLD ago. We can add push-based mechanisms later.
+- We include message data directly on invites.
+- Every state update is a DB write. This allows us to resume on cold start.
+- On cold start, we can scan the DB for UNINITIALIZED messages and payloads and resume sending them.
+- Repeated sends of the same payload should be idempotent. When receiving a message or invite, the receiving side will store the hash of the encrypted payload alongside the decrypted result. If a message is received with an id that already exists in the DB, it is ignored.
+- We have ignored race conditions for now (as network requests may take different amounts of time).
+  - It may make sense to later add multi-producer, single-consumer queues for (processMessages + refreshUserInstallations) and (processPayloads).
+  - Even simpler, we could have singleton threads for processMessages() and processPayloads() that simply run on an interval.
+
+### Creating a conversation
+
+```
+createConversation():
+    Deterministically derive convo_id from peer_address
+    If conversation with convo_id doesn't already exist in DB, insert it with conversation.state = UNINITIALIZED
+    return conversation
+```
+
+### Sending a message in a conversation
+
+```
+sendMessage():
+    Insert message into DB with message.state = UNINITIALIZED and message.convo_id set
+    processMessages() // Could be kicked off asynchronously or synchronously
+    return success
+
+processMessages():
+    For each message in UNINITIALIZED state, processing in order of timestamp:
+        Fetch the conversation's users from the DB (including self)
+        For each user:
+            If user.last_refreshed is uninitialized or more than THRESHOLD ago:
+                refreshUserInstallations()    // Could be kicked off asynchronously or synchronously
+                return  // refreshUserInstallations() will call back into processMessages() when ready
+        Fetch the installations of both users from the DB
+        For each installation:
+            // Build the plaintext payload
+            If conversation.state == UNINITIALIZED:
+                Construct the payload as an invite with the message attached
+            Else if conversation.state == INVITED:
+                Construct the payload as a message
+            // Encrypt the payload
+            If installation.state == UNINITIALIZED:
+                Create an outbound session (hold it in memory)
```

Comment:
What should happen in the case there is no session for channel but there is a conversation?